adding multi tenancy for controller using aad-pod-identity

nader-ziada opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](../CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**

<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
/kind other
-->

/kind feature

**What this PR does / why we need it**:
- adding azure principal and identity provider types for multi-tenancy

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #586 , #778   

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._
- work in progress

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
support for controller manager multi-tenancy 
```

nader-ziada commented:
@devigned I thought I would push this while I'm working on it still so you can take a look and let me know what you think of the direction

devigned commented on cloud/scope/clients.go:
```diff
@@ -82,3 +83,32 @@ func (c *AzureClients) setCredentials(subscriptionID string) error {
+	c.EnvironmentSettings = settings
+	c.ResourceManagerEndpoint = settings.Environment.ResourceManagerEndpoint
+	c.ResourceManagerVMDNSSuffix = settings.Environment.ResourceManagerVMDNSSuffix
+	c.Values[auth.ClientID] = strings.TrimSuffix(c.Values[auth.ClientID], "\n")
```
I think the only `c.Values` needed to be set here is just the SubscriptionID. The Authorizer should encapsulate the rest of this information.

devigned commented on api/v1alpha3/azurecluster_types.go:
```diff
@@ -48,6 +59,10 @@ type AzureClusterSpec struct {
+
+	// PrincipalRef is a reference to a principal to be used when reconciling this cluster
+	// +optional
+	PrincipalRef *AzurePrincipalRef `json:"principalRef,omitempty"`
```
I think we can simplify this to be the name of the identity to use for this cluster. The identity must reside inside of the cluster namespace and should be owned by the cluster.

All references should be to the known type of AzureIdentity via AAD Pod Identity.

devigned commented on api/v1alpha3/azureserviceprincipal_types.go:
```diff
@@ -0,0 +1,72 @@
+
+// AzureClusterPrincipal represents a reference to an Azure access key ID and
+// secret access key, stored in a secret.
+type AzureClusterPrincipal struct {
```
With the simplified design, I think this can be removed.

devigned commented on cloud/scope/principal.go:
```diff
@@ -0,0 +1,167 @@
+}
+
+// GetAuthorizer returnes Azure authorizer based on the provided principal
+func (p *AzureCredentialsProvider) GetAuthorizer(ctx context.Context) (autorest.Authorizer, error) {
```
Creating the Authorizer should be trivial using AAD pod identity. We can use the MSI endpoint provided. See the demo code here for an example of creating the Authorizer.

https://github.com/Azure/aad-pod-identity/blob/fb1bdba48ab5c1fa0be64382b6a74acdbcb41089/cmd/demo/main.go#L48

devigned commented on config/manager/manager_image_patch.yaml:
```diff
@@ -8,5 +8,5 @@ spec:
       containers:
       # Change the value of image field below to your controller image URL
-      - image: gcr.io/k8s-staging-cluster-api-azure/cluster-api-azure-controller:latest
+      - image: gcr.io/cf-london-servces-k8s/nziada/cluster-api-azure-controller-amd64:dev
```
oops... We need to stop manipulating files in place like this. It's too easy to check these in.

cc @CecileRobertMichon

CecileRobertMichon replied:
I actually tried to add this to gitignore last week and it didn't work (??) i don't know if I was doing it wrong or what kind of dark magic was going on...

nader-ziada replied:
I always check this one in!

devigned reviewed (changes requested):
Per our conversation in Slack: https://kubernetes.slack.com/archives/CEX9HENG7/p1602089427042600. I believe we are able to greatly simplify the design after talking with the AAD pod identity team. This is a great start down the multi-tenancy path.

The gist of the linked conversation:
> We should run aad pod identity within the capz namespace in namespace constrained mode. We should allow users to create `AzureIdentities` within the workload cluster namespace and have the identity owned by the `AzureCluster`. The capz controller would be responsible for copying the cluster namespace identity into the controller namespace and creating a `AzureIdentityBinding` in the controller namespace to link the label of the capz controller and the capz namespaced copy of the cluster `AzureIdentity`. The capz controller would then be responsible for deleting the identities within the capz namespace when they are no longer needed.

devigned commented:
Probably goes without saying, but we will also need to deploy aad pod identity components with the CAPZ components.

nader-ziada commented:
> Probably goes without saying, but we will also need to deploy aad pod identity components with the capz components.

yes, will add the deployment for it, was still trying to get it to work

CecileRobertMichon commented on Makefile:
```diff
@@ -438,6 +439,9 @@ create-management-cluster: $(KUSTOMIZE) $(ENVSUBST)
 	@echo 'Set kubectl context to the kind management cluster by running "kubectl config set-context kind-capz"'
 
+	# aad-pod-identity deploylment for handling identities 
+	$(ENVSUBST) < $(TEMPLATES_DIR)/addons/aad-pod-identity-deployment.yaml | kubectl apply -f -
```
how will this work for other ways of creating Azure clusters (Tilt, clusterctl, kustomize, etc.)? should we consider adding it to infrastructure-components?

nader-ziada replied:
Currently in the process of getting things to work, was going to add it to the tiltfile, didn’t think about infrastructure components

nader-ziada commented on cloud/scope/identity.go:
```diff
@@ -0,0 +1,192 @@
+	} else if p.Identity.Spec.Type == aadpodid.UserAssignedMSI {
+		spt, err = adal.NewServicePrincipalTokenFromMSIWithUserAssignedID(msiEndpoint, resourceManagerEndpoint, p.Identity.Spec.ClientID)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get token from user-assigned identity")
```
getting this error here so far
```
E1015 15:42:31.698604       1 mic.go:945] ignoring azure identity capz-system/user-assigned-id-nz1, error: invalid resource id: "", must match /subscriptions/<subid>/resourcegroups/<resourcegroup>/providers/Microsoft.ManagedIdentity/userAssignedIdentities/<name>
```

nader-ziada commented on cloud/scope/identity.go:
```diff
@@ -0,0 +1,192 @@
+
+	azureIdentity := aadpodid.AzureIdentity{}
+	err = crdClient.Get().Namespace(namespace).Resource(aadpodid.AzureIDResource).Name(identityName).Do().Into(&azureIdentity)
+	if err != nil {
```
This get doesn't seem to work properly

CecileRobertMichon commented on Makefile:
```diff
@@ -426,6 +426,7 @@ create-management-cluster: $(KUSTOMIZE) $(ENVSUBST)
 	kubectl wait --for=condition=Available --timeout=5m -n capi-kubeadm-bootstrap-system deployment -l cluster.x-k8s.io/provider=bootstrap-kubeadm
 	kubectl wait --for=condition=Available --timeout=5m -n capi-kubeadm-control-plane-system deployment -l cluster.x-k8s.io/provider=control-plane-kubeadm
 
+	sleep 10
```
you shouldn't need this anymore since #995

nader-ziada commented:
@devigned I updated the code to support SP identity as we discussed, please take a look 

I'm still working on docs and e2e test

nader-ziada commented:
@CecileRobertMichon @devigned can you please take a look, it ready for an initial review. I'm still working on adding an e2e test, but creating an identity first in azure using the sdk is not straight forward

devigned commented:
@nader-ziada what if we pre-provision the identity and keep it around to be used in tests? Might be easier than creating one each time.

devigned commented on docs/book/src/topics/multitenancy.md:
````diff
@@ -0,0 +1,57 @@
+  identityName: <name-of-identity>
+```  
+
+for more details on how aad-pod-identity work, please check the guide [here](https://azure.github.io/aad-pod-identity/docs/)
````
```suggestion
for more details on how aad-pod-identity works, please check the guide [here](https://azure.github.io/aad-pod-identity/docs/)
```

devigned commented on templates/flavors/multi-tenancy/kustomization.yaml:
```diff
@@ -0,0 +1,6 @@
+  - ../default
+  - azureidentity.yaml
+patchesStrategicMerge:
+  - patches/sp-multitenancy-identity.yaml
```
newline

devigned commented on templates/flavors/multi-tenancy/patches/sp-multitenancy-identity.yaml:
```diff
@@ -0,0 +1,7 @@
+metadata:
+  name: ${CLUSTER_NAME}
+spec:
+  identityName: ${CLUSTER_IDENTITY_NAME}
```
newline

devigned reviewed (changes requested):
Great work, @nader-ziada. The biggest question marks I have are related to the hardcoded namespace for the AzureIdentities and Bindings, and the validation that the IdentityName is static.

Looks really close! 🚀

nader-ziada commented on api/v1alpha3/azurecluster_validation.go:
```diff
@@ -54,10 +55,27 @@ func (c *AzureCluster) validateCluster() error {
+func (c *AzureCluster) validateIdentity(old *AzureCluster, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if to.String(c.Spec.IdentityName) != to.String(old.Spec.IdentityName) {
+		allErrs = append(allErrs, field.Invalid(fldPath, c.Spec.IdentityName, "IdentityName should not be modified after AzureCluster creation."))
```
if the VMs are created with a specific identity, what would happen if that changed, the new one will just replace the existing one? or will the NMI just use the new one going forward?

devigned replied:
The identity used to provision the infrastructure (`Cluster.Spec.IdentityName`) can change between each reconcile loop as long as it had the RBAC rights to mutate the infrastructure. This is the identity the CAPZ controller will assume via AAD Pod Identity to make REST requests to Azure to build / manipulate workload cluster infrastructure.

The identity used to provision the infrastructure is not the System Assigned or User Assigned Identity on provisioned workload machine.

I'd expect the new identity to be added to the controller (identity / binding moved into the controller namespace), the old IdentityBinding, possibly Identity, to be deleted from the controller namespace, and the next reconcile loop for the cluster resources to use a new Authorizer created using the new Identity ClientID.

CecileRobertMichon replied:
That makes sense to me, especially for cases where the identity needs to be rotated

nader-ziada replied:
ok, will remove this validation, but that means a new reconciler to remove the identities no longer used

nader-ziada replied:
removed validation and added reconciler

nader-ziada commented on cloud/scope/identity.go:
```diff
@@ -0,0 +1,127 @@
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      p.Identity.Name,
+			Namespace: "capz-system",
```
its hard-coded in the deployment yamls in the config folder, but I agree we should remove the hard-coding, will fix that

nader-ziada replied:
I think it has to be capz-system

CecileRobertMichon replied:
Is that documented in the CAPI provider contract? If not it's probably not safe to assume

CecileRobertMichon commented:
> but creating an identity first in azure using the sdk is not straight forward

you could do something like https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/974/files#diff-a18d58ea4b72ff83f83e62ae5a6500ea0de4759cdf9da9cdefa97b3b728d03b6R122

CecileRobertMichon commented on api/v1alpha3/azurecluster_types.go:
```diff
@@ -48,6 +51,10 @@ type AzureClusterSpec struct {
+
+	// IdentityName is a reference to a AzureIdentity to be used when reconciling this cluster
+	// +optional
+	IdentityName *string `json:"identityName,omitempty"`
```
This API change doesn't match what was proposed to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/docs/proposals/20200720-single-controller-multitenancy.md#cluster-api-provider-azure-v1alpha3-types 

If changes to the design had to be made, let's first update the proposal so it matches the implementation.

nader-ziada replied:
@devigned you mentioned you update the proposal? or I can do it if you haven't already changed it

devigned replied:
Will do. Just been short on time.

CecileRobertMichon replied:
let's make sure we update the proposal before merging this PR 

/hold

nader-ziada replied:
also wondering if we need a change to clusterctl move before merging this in

CecileRobertMichon replied:
is the identity owned by the cluster? if so it should be moved with it already

should be pretty easy to validate that assumption

nader-ziada replied:
I don't think its owned by the cluster, need to double check that

CecileRobertMichon commented on controllers/azurecluster_reconciler.go:
```diff
@@ -138,7 +141,41 @@ func (r *azureClusterReconciler) Delete(ctx context.Context) error {
+			binding.ObjectMeta.Labels[infrav1.ClusterLabelNamespace] == r.scope.AzureCluster.Namespace {
+
+			identityNamesToDelete = append(identityNamesToDelete, binding.Spec.AzureIdentity)
+			if err := r.scope.Client.Delete(ctx, &binding); err != nil {
```
I might be misreading the code, but, are we deleting the identity that wasn't created by CAPZ? or is it created somewhere else? 

If it's applied as part of the deployment, would it make sense to have this in a separate reconciler since it's not technically part of the AzureCluster resource?

nader-ziada replied:
so we create a copy of the azureidentity in the capz-system namespace so the NMI pod can use it, and that's the one we delete. its created as part of the initialization of the cluster scope 

do you think it needs a different reconciler to remove it if the cluster is deleted? I had that at some point, but removed it

nader-ziada replied:
we now have azureidentity_controller

CecileRobertMichon commented on controllers/azurecluster_reconciler.go:
```diff
@@ -138,7 +141,41 @@ func (r *azureClusterReconciler) Delete(ctx context.Context) error {
+	}
+	for _, identityName := range identityNamesToDelete {
+		azureIdentity := &aadpodv1.AzureIdentity{}
+		if err := r.scope.Client.Get(ctx, client.ObjectKey{Name: identityName, Namespace: "capz-system"}, azureIdentity); err != nil {
```
`capz-system` should be a const reused across the different places that have it hardcoded

nader-ziada replied:
its hardcoded in the yaml, not in the code, I'm removing the hard-coding in the code, but don't know how to do that with yaml https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/f9da6a5e82da5ccd7c692f47704015be27f7ba6e/config/default/kustomization.yaml#L1

nader-ziada commented:
> > but creating an identity first in azure using the sdk is not straight forward
> 
> you could do something like https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/974/files#diff-a18d58ea4b72ff83f83e62ae5a6500ea0de4759cdf9da9cdefa97b3b728d03b6R122

I was thinking I would create the identity in the e2e script using `az ad sp create-for-rbac`. Is that reasonable? or I can change it if you prefer, but there is no straight forward azure sdk api to create a service principal  https://github.com/Azure/azure-sdk-for-go/issues/596

CecileRobertMichon commented:
@nader-ziada what about https://github.com/Azure-Samples/azure-sdk-for-go-samples/blob/master/graphrbac/graph.go#L43 (from https://github.com/Azure/azure-sdk-for-go/issues/596#issuecomment-392813961) ? using CLI commands is okay as last resort but doing this programmatically would be preferable IMO

nader-ziada commented:
> @nader-ziada what about https://github.com/Azure-Samples/azure-sdk-for-go-samples/blob/master/graphrbac/graph.go#L43 (from [Azure/azure-sdk-for-go#596 (comment)](https://github.com/Azure/azure-sdk-for-go/issues/596#issuecomment-392813961)) ? using CLI commands is okay as last resort but doing this programmatically would be preferable IMO

ok, let me try that

nader-ziada commented:
still have to update e2e tests and remove the validation about updating the identity on cluster

nader-ziada commented:
@CecileRobertMichon @devigned I think I addressed all the comments, but had to take out the e2e test I already had for now because the ci user doesn't have permission to create an AAD identity `Insufficient privileges to complete the operation`.
`

nader-ziada commented:
also not sure if we need to wait for a change with clusterctl move?

devigned commented on test/e2e/azure_test.go:
```diff
@@ -350,4 +350,5 @@ var _ = Describe("Workload cluster creation", func() {
 			})
 		})
 	})
+
```
Looks like a stray extra newline.

nader-ziada replied:
removed

devigned reviewed (changes requested):
The identity reconciler looks great! 

I have one superficial change request, and one left over from last review: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/977/files#r516679923

Looks great!

nader-ziada commented on cloud/scope/identity.go:
```diff
@@ -0,0 +1,127 @@
+		},
+		Spec: aadpodv1.AzureIdentityBindingSpec{
+			AzureIdentity: copiedIdentity.Name,
+			Selector:      "capz-controller-aadpodidentity-selector", //should be same as selector added on controller
```
done, sorry missed this one

CecileRobertMichon commented on cloud/defaults.go:
```diff
@@ -39,6 +39,16 @@ const (
 )
 
+// GenerateInternalLBName generates a internal load balancer name, based on the cluster name.
+func GenerateInternalLBName(clusterName string) string {
```
why is this needed? It should have been removed since #974

nader-ziada replied:
must have missed something with the rebase, fixed it now

CecileRobertMichon commented on cloud/defaults.go:
```diff
@@ -39,6 +39,16 @@ const (
+}
+
+// GeneratePublicLBName generates a public load balancer name, based on the cluster name.
+func GeneratePublicLBName(clusterName string) string {
```
same question here

nader-ziada replied:
same

CecileRobertMichon commented on config/aadpodidentity/aad-pod-identity-deployment.yaml:
```diff
@@ -0,0 +1,158 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```
@aramase can you please take a look to make sure everything looks good from an AAD pod identity standpoint?

aramase replied:
Thank you for the ping. I'll take a look at it today.

aramase commented:
/cc

aramase commented on cloud/scope/identity.go:
```diff
@@ -0,0 +1,129 @@
+	}
+
+	azureIdentity := &aadpodv1.AzureIdentity{}
+	err := kubeClient.Get(ctx, client.ObjectKey{Name: identityName, Namespace: namespace}, azureIdentity)
```
If the `identityName` is configured, then it's only applicable for pod identity? Can they configure managed identity without pod identity?

nader-ziada replied:
yes this is only used for the pod identity, and its optional, users can already configure system assigned and user managed identity before this change

aramase commented on cloud/scope/identity.go:
```diff
@@ -0,0 +1,129 @@
+	}, nil
+}
+
+// GetAuthorizer returnes Azure authorizer based on the provided azure identity
```
nit: s/returnes/returns

nader-ziada replied:
> does pod identity get deployed by default with this PR or is it only deployed when there is an `AzureCluster` with `identityName`? If deployed by default, how would we handle scenarios where the cluster already has pod identity running in `standard` mode?

yes, its installed on the management cluster by default in the same namespace as the manager controller, what is a use case where we would have the standard mode also running?

aramase commented on cloud/scope/identity.go:
```diff
@@ -0,0 +1,129 @@
+				clusterv1.ClusterLabelName:    p.AzureCluster.Name,
+				infrav1.ClusterLabelNamespace: p.AzureCluster.Namespace,
+			},
+			OwnerReferences: p.AzureCluster.OwnerReferences,
```
+1

aramase commented on cloud/scope/identity.go:
```diff
@@ -0,0 +1,129 @@
+			APIVersion: "aadpodidentity.k8s.io/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      p.Identity.Name,
```
How do we handle `AzureIdentity` with same name in multi namespaces? When run in namespaced mode, user could configure `azid1` in `ns1` and `azid1` in `ns2` (both with different `clientIDs`).

cluster1 referring to `ns1/azid1` and cluster referring to `ns2/azid` will result in conflict while copied in controller namespace? Should we add the namespace to the copied `AzureIdentity` name?

nader-ziada replied:
good idea, will add the namespace to the name

aramase commented on config/aadpodidentity/aad-pod-identity-deployment.yaml:
```diff
@@ -0,0 +1,158 @@
+  group: aadpodidentity.k8s.io
+  version: v1
+  names:
+    kind: AzureAssignedIdentity
```
`AzureAssignedIdentity` crd isn't required for managed mode. 

The deploy yaml for managed mode: https://github.com/Azure/aad-pod-identity/blob/v1.6.3/deploy/infra/managed-mode-deployment.yaml

nader-ziada replied:
will remove

aramase commented on config/rbac/role.yaml:
```diff
@@ -29,6 +29,44 @@ rules:
+- apiGroups:
+  - aadpodidentity.k8s.io
+  resources:
+  - azureassignedidentities
```
`azureassignedidentities` are not required for managed mode

nader-ziada replied:
will remove as well

aramase commented on controllers/azureidentity_controller.go:
```diff
@@ -0,0 +1,167 @@
+}
+
+// +kubebuilder:rbac:groups=aadpodidentity.k8s.io,resources=azureassignedidentities,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=aadpodidentity.k8s.io,resources=azureassignedidentities/status,verbs=get;update;patch
```
Not required for managed mode

aramase commented on docs/book/src/topics/multitenancy.md:
```diff
@@ -0,0 +1,57 @@
+
+## Service Principal Identity
+
+Once a new SP Identity is created in Azure, a new resource of AzureIdentity should be created in the managment cluster, for example
```
Might be good to add a reference to this doc - https://azure.github.io/aad-pod-identity/docs/concepts/azureidentity/

aramase commented on main.go:
```diff
@@ -73,6 +76,20 @@ func init() {
+		&aadpodv1.AzureIdentityList{},
+		&aadpodv1.AzureIdentityBinding{},
+		&aadpodv1.AzureIdentityBindingList{},
+		&aadpodv1.AzureAssignedIdentity{},
```
Not required for managed mode

aramase reviewed (commented):
Added few comments.

Also, does pod identity get deployed by default with this PR or is it only deployed when there is an `AzureCluster` with `identityName`? If deployed by default, how would we handle scenarios where the cluster already has pod identity running?

nader-ziada commented:
> does pod identity get deployed by default with this PR or is it only deployed when there is an AzureCluster with identityName? If deployed by default, how would we handle scenarios where the cluster already has pod identity running in standard mode?

@aramase  yes, its installed on the management cluster by default in the same namespace as the manager controller, what is a use case where we would have the standard mode also running?

aramase commented:
> @aramase yes, its installed on the management cluster by default in the same namespace as the manager controller, what is a use case where we would have the standard mode also running?

@nader-ziada It's possible the user is using pod identity in standard/managed mode as part of their workload deployments. If they want to also use the cluster as mgmt cluster, then capz controller will try to deploy pod identity again in managed mode in `capz-system` namespace. This would mean the NMI pods in `capz-system` namespace will fail to run because of host port conflicts. This will not actually affect the other functioning NMI pods the user has deployed in their own/kube-system namespace. 

What if we do provide user's the option to opt out from installing pod identity as part of the capz controller and instead just manage it on their own? By default we can still continue it install it.

nader-ziada commented:
> > @aramase yes, its installed on the management cluster by default in the same namespace as the manager controller, what is a use case where we would have the standard mode also running?
> 
> @nader-ziada It's possible the user is using pod identity in standard/managed mode as part of their workload deployments. If they want to also use the cluster as mgmt cluster, then capz controller will try to deploy pod identity again in managed mode in `capz-system` namespace. This would mean the NMI pods in `capz-system` namespace will fail to run because of host port conflicts. This will not actually affect the other functioning NMI pods the user has deployed in their own/kube-system namespace.
> 
> What if we do provide user's the option to opt out from installing pod identity as part of the capz controller and instead just manage it on their own? By default we can still continue it install it.

I'm not sure how we can do that with kustomize, I'll look into if we can only deploy a `deployment` based on flag 
cc @devigned

nader-ziada commented:
/test pull-cluster-api-provider-azure-e2e-full

nader-ziada commented:
@CecileRobertMichon @devigned 
The only two things outstanding here are:

 - What do we want to do if a users doesn't wants to deploy aad-pod-identity as part of capz
```
@aramase yes, its installed on the management cluster by default in the same namespace as the manager controller, what is a use case where we would have the standard mode also running?

@nader-ziada It's possible the user is using pod identity in standard/managed mode as part of their workload deployments. If they want to also use the cluster as mgmt cluster, then capz controller will try to deploy pod identity again in managed mode in capz-system namespace. This would mean the NMI pods in capz-system namespace will fail to run because of host port conflicts. This will not actually affect the other functioning NMI pods the user has deployed in their own/kube-system namespace.

What if we do provide user's the option to opt out from installing pod identity as part of the capz controller and instead just manage it on their own? By default we can still continue it install it.

I'm not sure how we can do that with kustomize, I'll look into if we can only deploy a deployment based on flag
cc @devigned
```
- what is supposed to happen with clusterctl move, its currently not moving the AzureIdentity and AzureIdentityBinding in the capz-system namespace even though they have the cluster as an OwnerReference and the clusterctl move label
https://github.com/kubernetes-sigs/cluster-api/issues/3933

nader-ziada commented:
> * what is supposed to happen with clusterctl move, its currently not moving the AzureIdentity and AzureIdentityBinding in the capz-system namespace even though they have the cluster as an OwnerReference and the clusterctl move label
>   [kubernetes-sigs/cluster-api#3933](https://github.com/kubernetes-sigs/cluster-api/issues/3933)

The issue with this is not just the move, but its because we expect the user to create an AzureIdentity with the cluster and we copy it over to the capz-system namespace so the nmi pod can use it, but we can't move that user created identity since its not in clusterapi control

nader-ziada commented:
@devigned @CecileRobertMichon added the custom type to azurecluster to deal with with the `move` command as we discussed, PTAL

devigned commented on api/v1alpha3/types.go:
```diff
@@ -321,6 +323,42 @@ type UserAssignedIdentity struct {
+	ClientID string `json:"clientID"`
+	// Used for service principal
+	// +optional
+	ClientPassword corev1.SecretReference `json:"clientPassword,omitempty"`
```
```suggestion
	ClientSecret corev1.SecretReference `json:"clientSecret,omitempty"`
```

These are usually referred to as "Client Secrets" in UI and doc https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal#option-2-create-a-new-application-secret. Using password, might make it seem like a only the text secret is usable rather than certificate secret as well.

devigned commented on api/v1alpha3/types.go:
```diff
@@ -321,6 +323,42 @@ type UserAssignedIdentity struct {
+	ResourceID string `json:"resourceID,omitempty"`
+	// Both User Assigned MSI and SP can use this field.
+	ClientID string `json:"clientID"`
+	// Used for service principal
```
```suggestion
	// ClientSecret is a secret reference which should contain either a Service Principal password or certificate secret.
```

devigned commented on cloud/scope/identity.go:
```diff
@@ -0,0 +1,149 @@
+	}
+
+	if azureCluster.Spec.IdentitySpec.Type != infrav1.ServicePrincipal {
+		return nil, errors.New("AzureIdentity is not of type Service Principal")
```
WDYT about a web hook validation for this?

devigned commented on templates/flavors/multi-tenancy/patches/sp-multitenancy-identity.yaml:
```diff
@@ -0,0 +1,13 @@
+    tenantID: ${AZURE_TENANT_ID}
+    clientID: ${AZURE_CLUSTER_IDENTITY_CLIENT_ID}
+    clientPassword: {"name":"${AZURE_CLUSTER_IDENTITY_SECRET_NAME}","namespace":"${AZURE_CLUSTER_IDENTITY_SECRET_NAMESPACE}"}
+  
```
new line

devigned commented on test/e2e/azure_identity.go:
```diff
@@ -0,0 +1,66 @@
+// an AzureCluster and used as the identity for that cluster.
+func AzureServicePrincipalIdentitySpec(ctx context.Context, inputGetter func() AzureServicePrincipalIdentitySpecInput) {
+	var (
+		specName = "azure-identity"
```
const?

nader-ziada replied:
this just the name of the spec

devigned commented on api/v1alpha3/types.go:
```diff
@@ -321,6 +323,42 @@ type UserAssignedIdentity struct {
+)
+
+// IdentitySpec specifies the parameters that are used to create an AzureIdentity
+type IdentitySpec struct {
```
We had discussed having `AllowedNamespaces` label selector on the `IdentitySpec`, similar to the design approach in CAPA: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/docs/proposal/20200506-single-controller-multitenancy.md#namespace-restrictions

This could be added in a subsequent PR.

nader-ziada replied:
I started adding it, will submit it with the rest of the comments

devigned reviewed (commented):
Looks great. I've added some small feedback items.

nader-ziada commented:
@devigned addressed all the comments except `allowed namespaces`, need to think about it a bit more

devigned commented on controllers/azurecluster_controller.go:
```diff
@@ -215,6 +217,23 @@ func (r *AzureClusterReconciler) reconcileNormal(ctx context.Context, clusterSco
+			if err != nil {
+				return reconcile.Result{}, errors.Wrap(err, "failed to init patch helper")
+			}
+			identity.ObjectMeta.OwnerReferences = azureCluster.GetOwnerReferences()
```
I'm wondering if this would be better placed in the cluster scope so that upon scope.Close() the patch will be ensured.

nader-ziada replied:
ok, makes sense, will move it

nader-ziada replied:
so the patch helper takes the azurecluster when its created, so it will not work with patching the identity

devigned commented on controllers/azureidentity_controller.go:
```diff
@@ -0,0 +1,166 @@
+
+// SetupWithManager initializes this controller with a manager
+func (r *AzureIdentityReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	log := r.Log.WithValues("controller", "AzureCluster")
```
I think there might be some copypasta here.

devigned commented on templates/flavors/multi-tenancy/azure-cluster-identity.yaml:
```diff
@@ -0,0 +1,10 @@
+  type: ServicePrincipal
+  tenantID: "${AZURE_TENANT_ID}"
+  clientID: "${AZURE_CLUSTER_IDENTITY_CLIENT_ID}"
+  clientSecret: {"name":"${AZURE_CLUSTER_IDENTITY_SECRET_NAME}","namespace":"${AZURE_CLUSTER_IDENTITY_SECRET_NAMESPACE}"}
```
new line

devigned reviewed (changes requested):
Mostly little nits. Looking good.

nader-ziada commented on api/v1alpha3/azureclusteridentity_types.go:
```diff
@@ -0,0 +1,98 @@
+type AzureClusterIdentityStatus struct {
+	// Ready is true when the provider resource is ready.
+	// +optional
+	Ready bool `json:"ready"`
```
not really,

nader-ziada commented:
/test pull-cluster-api-provider-azure-coverage

nader-ziada commented:
@devigned ready for another look

nader-ziada commented:
/test pull-cluster-api-provider-azure-e2e-full
/test pull-cluster-api-provider-azure-e2e

CecileRobertMichon commented on docs/book/src/topics/multitenancy.md:
```diff
@@ -0,0 +1,74 @@
+
+## Service Principal Identity
+
+Once a new SP Identity is created in Azure, the coresponding values should be used to create an `AzureClusterIdentity` resource
```
```suggestion
Once a new SP Identity is created in Azure, the corresponding values should be used to create an `AzureClusterIdentity` resource:
```

CecileRobertMichon commented on docs/book/src/topics/multitenancy.md:
````diff
@@ -0,0 +1,74 @@
+  password: PASSWORD
+```
+
+The Identity can be added to an `AzureCluster` by using `IdentityRef` field
````
```suggestion
The Identity can be added to an `AzureCluster` by using `IdentityRef` field:
```

CecileRobertMichon commented on docs/book/src/topics/multitenancy.md:
````diff
@@ -0,0 +1,74 @@
+    namespace: <namespace-of-identity>
+```
+
+for more details on how aad-pod-identity works, please check the guide [here](https://azure.github.io/aad-pod-identity/docs/)
````
```suggestion
For more details on how aad-pod-identity works, please check the guide [here](https://azure.github.io/aad-pod-identity/docs/).
```

CecileRobertMichon commented on controllers/helpers.go:
```diff
@@ -387,3 +387,20 @@ func ShouldDeleteIndividualResources(ctx context.Context, clusterScope *scope.Cl
 	return err != nil || !managed
 }
+
+// GetClusterIdentityFromRef returns the AzureClusterIdentity referenced by the AzureCluster
```
```suggestion
// GetClusterIdentityFromRef returns the AzureClusterIdentity referenced by the AzureCluster.
```

CecileRobertMichon commented on test/e2e/azure_test.go:
```diff
@@ -350,4 +351,60 @@ var _ = Describe("Workload cluster creation", func() {
 	})
+
+	Context("Creating a cluster using a different SP identity", func() {
+		BeforeEach(func() {
```
Same comment I wrote in the Windows PR - we need to audit all the clusters that get created in pr e2e tests and see if any of them can be safely combined to avoid growing the cost and duration of each PR run too much

nader-ziada replied:
I think it makes sense for this test to have its own cluster since it using a different identity which we would want to test separately, wdyt?

CecileRobertMichon commented on docs/book/src/topics/multitenancy.md:
```diff
@@ -0,0 +1,74 @@
+type: Opaque
+data:
+  certificate: CERTIFICATE
+  password: PASSWORD
```
this is unclear to me? is it one or the other?

nader-ziada replied:
both kinds are supported by aad-pod-identity https://github.com/Azure/aad-pod-identity/tree/master/deploy/demo

CecileRobertMichon commented on docs/book/src/topics/multitenancy.md:
````diff
@@ -0,0 +1,74 @@
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: AzureClusterIdentity
````
Should this document AllowedNamespaces?

CecileRobertMichon commented on controllers/azureidentity_controller.go:
```diff
@@ -0,0 +1,166 @@
+		return ctrl.Result{}, err
+	}
+
+	bindingsToDelete := []aadpodv1.AzureIdentityBinding{}
```
where are the bindings created?

CecileRobertMichon replied:
nevermind found it in scope. Seems a bit weird to me to not create the bindings/azureidentity in here too. Also why are we expecting unused bindings to be left around needing cleanup? Why not delete them as part of AzureCluster deletion?

nader-ziada replied:
The thinking behind this is that this object lives in a different namespace and is only for the purpose of the controller with the nmi pod, technically the cluster/user should not care about it. Also the reason we needed the cleanup controller is that if the user changes the reference to a different identity, the controller will generate a new azureIdentity/Binding and the old one will be left behind

CecileRobertMichon commented on cloud/scope/identity.go:
```diff
@@ -0,0 +1,162 @@
+			APIVersion: "aadpodidentity.k8s.io/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      identity.GetAzureIdentityName(p.AzureCluster.Name, p.AzureCluster.Namespace, p.Identity.Name),
```
if this has the cluster name in it, its lifecycle is owned by the life of the cluster, no? Why not create/delete it when the an AzureCluster is created/deleted, similarly to how we create/delete the azure.json secret?

nader-ziada replied:
I just added the cluster name to make it easier to find, but it wasn't there before, its not part of the cluster lifecycle it lives in a different namespace and is for internal use by the controller

nader-ziada commented on api/v1alpha3/azureclusteridentity_types.go:
```diff
@@ -0,0 +1,105 @@
+// AzureClusterIdentitySpec defines the parameters that are used to create an AzureIdentity
+type AzureClusterIdentitySpec struct {
+	// UserAssignedMSI or Service Principal
+	Type IdentityType `json:"type"`
```
added a kubebuilder validation ` // +kubebuilder:validation:Enum=None;SystemAssigned;UserAssigned`

wdyt?

nader-ziada commented:
/test pull-cluster-api-provider-azure-e2e-full

nader-ziada commented:
/test pull-cluster-api-provider-azure-e2e-full

CecileRobertMichon commented:
lgtm

Are you tracking all follow-ups (clusterctl move, pre-existing aad pod identity deploy, etc) in issues in there respective repos?

devigned reviewed (approved):
Solid work, @nader-ziada!
lgtm

devigned commented:
:rocket:
/lgtm
/approve

CecileRobertMichon commented:
Just remembered still waiting on the proposal to match the implementation

That's why there's a /hold

devigned commented:
@CecileRobertMichon I promise, promise, promise to update the proposal by eod tomorrow.

CecileRobertMichon commented:
/hold cancel 

@devigned thanks, github is witness 😉
